framework/src/org/checkerframework/framework/util/typeinference8/types/AbstractType.java

```diff
@@ -0,0 +1,343 @@
+        // TODO this isn't matching the JavaDoc.
+        return ((Type) getJavaType()).isParameterized();
+    }
+
+    /**
+     * @return the most specific array type that is a super type of this type or null if one doesn't
+     *     exist
+     */
+    public final AbstractType getMostSpecificArrayType() {
+        TypeMirror mostSpecific =
+                TypesUtils.getMostSpecificArrayType(getJavaType(), context.env.getTypeUtils());
+        if (mostSpecific != null) {
+            return create(mostSpecific);
+        } else {
+            return null;
+        }
+    }
+
+    /** @return true if this type is a primitive array. */
+    public final boolean isPrimitiveArray() {
+        return getJavaType().getKind() == TypeKind.ARRAY
+                && ((ArrayType) getJavaType()).getComponentType().getKind().isPrimitive();
+    }
+
+    /**
+     * @return assuming type is an intersection type, this method returns the bounds in this type
+     */
+    public final List<AbstractType> getIntersectionBounds() {
+        List<AbstractType> bounds = new ArrayList<>();
+        for (TypeMirror bound : ((IntersectionType) getJavaType()).getBounds()) {
+            bounds.add(create(bound));
+        }
+        return bounds;
+    }
+
+    /**
+     * @return assuming this type is a type variable, this method returns the upper bound of this
+     *     type.
+     */
+    public final AbstractType getTypeVarUpperBound() {
+        return create(((TypeVariable) getJavaType()).getUpperBound());
+    }
+    /**
+     * @return assuming this type is a type variable that has a lower bound, this method returns the
+     *     lower bound of this type.
+     */
+    public final AbstractType getTypeVarLowerBound() {
+        return create(((TypeVariable) getJavaType()).getLowerBound());
+    }
+
+    /** @return true if this type is a type variable with a lower bound */
+    public final boolean isLowerBoundTypeVariable() {
+        return ((TypeVariable) getJavaType()).getLowerBound().getKind() != TypeKind.NULL;
+    }
+
+    /**
+     * @return true if this type is a parameterized type whose has at least one wildcard as a type
+     *     argument.
+     */
+    public final boolean isWildcardParameterizedType() {
+        return TypesUtils.isWildcardParameterized(getJavaType());
+    }
+
+    /** @return this type's type arguments or null this type isn't a declared type */
+    public final List<AbstractType> getTypeArguments() {
+        if (getJavaType().getKind() != TypeKind.DECLARED) {
+            return null;
+        }
+        List<AbstractType> list = new ArrayList<>();
+        for (TypeMirror typeArg : ((DeclaredType) getJavaType()).getTypeArguments()) {
+            list.add(create(typeArg));
+        }
+        return list;
+    }
+
+    /** @return true if the type is an unbound wildcard */
+    public final boolean isUnboundWildcard() {
+        if (getJavaType().getKind() == TypeKind.WILDCARD) {
```

Comment:
Could you use the similarly-named methods in `TypesUtils` that this PR is also adding? Would reduce duplication a bit. Also, the logic there is slightly different.